### PR TITLE
Snapshot Testing

### PR DESCRIPTION
**Pull request by Kayanski**
This PR aims at introducing snapshot testing for migrations of smart-contracts. This is used to make sure that no breaking change to underlying state is introduced my version modifications : 

  - [x] Implementation
  - [x] Snapshot testing documentation

## Implementation

We introduce a new `snapshot_testing` method on the `Mock` structure. We advise using it at the end of every test that need stabilizing. It dumps the whole contract states of your structures and snapshot tests it.

You can test it by:
- commenting out the line `counter.increment()?;` in the `packages/cw-orch-mock/tests/conditional.rs` file
- running `cargo test`

More information about snapshot testing here
https://insta.rs/docs/quickstart/

### CODE REVIEW

File: packages/cw-orch-mock/Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,7 @@ cosmwasm-std = { workspace = true }
 cw-multi-test = { workspace = true }
 cw-utils = { workspace = true }
 serde = { workspace = true }
+insta = { version = "1.34.0", features = ["yaml"] }
```
Comment (CyberHoward) [inline packages/cw-orch-mock/Cargo.toml:18]: Can this be optional/feature flagged?

File: packages/cw-orch-mock/src/core.rs [hunk 2]
```diff
@@ -175,6 +175,22 @@ impl<S: StateInterface> Mock<S> {
...
     }
+
+    /// This functions allows for making sure the tests are stabilised and changes made to contracts don't have an impact on the internal storage
+    /// Usage:
+    /// ```rust
+    /// mock.snapshot_testing();
+    /// ```
+    /// Find more details on how snapshot testing works on the official quick-start guide: https://insta.rs/docs/quickstart/
+    /// This function will panic if the snapshot is different from the reference snapshot
+    pub fn snapshot_testing(&self) -> Result<(), CwEnvError> {
```
Comment (CyberHoward) [inline packages/cw-orch-mock/src/core.rs:186]: Can we rename this to `take_snapshot`?

File: packages/cw-orch-mock/src/snapshots/cw_orch_mock__core__snapshot_testing.snap
Comment (CyberHoward) [inline packages/cw-orch-mock/src/snapshots/cw_orch_mock__core__snapshot_testing.snap:1]: Could you add these paths to the `.gitattributes` as generated files?

**Comment by AndrewPochapsky** [comment]
This is great! I never knew that I wanted snapshot testing in cosmwasm but it seems like the perfect fit :)

File: docs/src/integrations/cw-multi-test.md [hunk 3]
```diff
@@ -25,25 +25,42 @@ You can then use the resulting `Mock` variable to interact with your [contracts]
...
+    let mock = Mock::new(Addr::unchecked("sender"));
+
+    ... // Upload, instantiate, execute contracts
+
+    // Make sure that the operations have a fixed result
+    mock.take_snapshot()?;
+}
+```
+
+At any point of development, if the storage variables are modified, this test will fail and alert you that you are doing breaking changes to your storage variables. Learn more about thi underlying tool on the [official documentation](https://insta.rs/).
```
Comment (CyberHoward) [inline docs/src/integrations/cw-multi-test.md:58]:
```suggestion
At any point of development, if the storage variables are modified, this test will fail and alert you that you are doing breaking changes to your storage variables. Learn more about the underlying tool in the [official documentation](https://insta.rs/).
```

File: packages/cw-orch-mock/src/core.rs [hunk 4]
```diff
@@ -175,6 +175,23 @@ impl<S: StateInterface> Mock<S> {
...
+
+    #[cfg(feature = "snapshot-testing")]
+    /// This functions allows for making sure the tests are stabilised and changes made to contracts don't have an impact on the internal storage
+    /// Usage:
+    /// ```rust,ignore
+    /// mock.take_snapshot();
+    /// ```
+    /// Find more details on how snapshot testing works on the official quick-start guide: https://insta.rs/docs/quickstart/
+    /// This function will panic if the snapshot is different from the reference snapshot
+    pub fn take_snapshot(&self) -> Result<(), CwEnvError> {
```
Comment (CyberHoward) [inline packages/cw-orch-mock/src/core.rs:187]: Should we allow for an optional snapshot name here? 

Can be provided as the first argument to the macro:
https://docs.rs/insta/1.34.0/insta/macro.assert_yaml_snapshot.html

**Review by CyberHoward — APPROVED** [review]

**Comment by Kayanski** [comment]
Today, the take_storage_snapshot function is a member of the Mock and thus the snapshots have to be named to be recognized from one another.
We could prevent that by changing that function to a macro, which would allow to not input a name for the snaps. 
However, that would remove the function from the Mock `API` and thus would make it harder to find. What do you think ?

**Comment by CyberHoward** [comment]
It's good that it needs a name but in the code it doesn't?

**Comment by Kayanski** [comment]
> It's good that it needs a name but in the code it doesn't? See my comment above.

Oh yeah I have added a name after you suggested it

File: docs/src/integrations/cw-multi-test.md [hunk 5]
```diff
@@ -25,25 +25,42 @@ You can then use the resulting `Mock` variable to interact with your [contracts]
...
+```rust,ignore
+
+#[test]
+fn storage_stays_thtake_storage_snapshot
+    let mock = Mock::new(Addr::unchecked("sender"));
+
+    ... // Upload, instantiate, execute contracts
+
+    // Make sure that the operations have a fixed result
+    mock.take_storage_snapshot()?;
```
Comment (CyberHoward) [inline docs/src/integrations/cw-multi-test.md:54]: Provide snapshot name here

**Pull request merged**